### ISSUE TITLE
Manifest failure: Can't read manifest: equal mark expected

### ISSUE THREAD

**Issue by zavalyshyn** [oscarlab/graphene#198]
Hi,

I'm trying to port Spring Framework into Graphene. More specifically, the hello world example of [REST API File Storage Server](https://github.com/spring-guides/gs-uploading-files/tree/master/complete). I can successfully build it for Graphene SGX but then it fails to run, giving me the "Can't read manifest: equal mark expected" error. 

I have my spring.manifest.template, in which I just add all the files from the project as trusted

**spring.manifest.template**
````
#!$(PAL)

loader.preload = file:$(SHIMPATH)
loader.exec = file:spring-rest/gradlew
loader.env.LD_LIBRARY_PATH = /lib:/lib/x86_64-linux-gnu:/spring-rest
loader.debug_type = none

fs.mount.lib.type = chroot
fs.mount.lib.path = /lib
fs.mount.lib.uri = file:$(LIBCDIR)

fs.mount.lib64.type = chroot
fs.mount.lib64.path = /lib/x86_64-linux-gnu
fs.mount.lib64.uri = file:/lib/x86_64-linux-gnu

fs.mount.build.type = chroot
fs.mount.build.path = /spring-rest
fs.mount.build.uri = file:spring-rest

fs.mount.usr.type = chroot
fs.mount.usr.path = /usr
fs.mount.usr.uri = file:/usr

fs.mount.tmp.type = chroot
fs.mount.tmp.path = /tmp
fs.mount.tmp.uri = file:/tmp

# allow to bind on port for listening
net.allow_bind.1 = $(HOST):$(PORT)

sys.stack.size = 256K
sys.brk.size = 4M
glibc.heap_size = 16M

sgx.thread_num = 28

sgx.trusted_files.ld = file:$(LIBCDIR)/ld-linux-x86-64.so.2
sgx.trusted_files.libc = file:$(LIBCDIR)/libc.so.6
sgx.trusted_files.libdl = file:$(LIBCDIR)/libdl.so.2
sgx.trusted_files.libm = file:$(LIBCDIR)/libm.so.6
sgx.trusted_files.libpthread = file:$(LIBCDIR)/libpthread.so.0
sgx.trusted_files.libssl = file:/lib/x86_64-linux-gnu/libssl.so.1.0.0
sgx.trusted_files.libcrypto = file:/lib/x86_64-linux-gnu/libcrypto.so.1.0.0

sgx.trusted_files.gradle1 = file:spring-rest/gradle/wrapper/gradle-wrapper.jar
sgx.trusted_files.gradle2 = file:spring-rest/gradle/wrapper/gradle-wrapper.properties
sgx.trusted_files.gradle3 = file:spring-rest/build.gradle

sgx.trusted_files.java1 = file:spring-rest/src/main/java/hello/Application.java
sgx.trusted_files.java2 = file:spring-rest/src/main/java/hello/FileUploadController.java
sgx.trusted_files.java3 = file:spring-rest/src/main/java/hello/storage/FileSystemStorageService.java
sgx.trusted_files.java4 = file:spring-rest/src/main/java/hello/storage/StorageException.java
sgx.trusted_files.java5 = file:spring-rest/src/main/java/hello/storage/StorageFileNotFoundException.java
sgx.trusted_files.java6 = file:spring-rest/src/main/java/hello/storage/StorageProperties.java
sgx.trusted_files.java7 = file:spring-rest/src/main/java/hello/storage/StorageService.java

sgx.trusted_files.resources1 = file:spring-rest/src/main/resources/application.properties
sgx.trusted_files.resources2 = file:spring-rest/src/main/resources/templates/uploadForm.html

sgx.trusted_files.test-java1 = file:spring-rest/src/test/java/hello/FileUploadIntegrationTests.java
sgx.trusted_files.test-java2 = file:spring-rest/src/test/java/hello/FileUploadTests.java
sgx.trusted_files.test-java3 = file:spring-rest/src/test/java/hello/storage/FileSystemStorageServiceTests.java

sgx.trusted_files.test-resources1 = file:spring-rest/src/test/resources/hello/testupload.txt
````

I have my Makefile that builds Spring Boot and tries to run it

**Makefile**
```Makefile
manifests = spring.manifest
SRCDIR = spring-rest
HOST = $(firstword $(shell ifconfig | grep 'inet addr:'| grep -v '127.0.0.1' -m 1 | cut -d: -f2))
PORT = 8090
CORES = 4
THREADS = 25


target = build-spring
exec_target = $(manifests)

extra_rules = -e 's:\$$(HOST):$(HOST):g' -e 's:\$$(PORT):$(PORT):g'

level = ../../
include ../../Makefile

.PHONY: build-spring
build-spring: spring-rest/gradlew

.PHONY: spring-rest/gradlew
spring-rest/gradlew: $(SRCDIR)
        cd $(SRCDIR) && ./gradlew build

.PHONY: $(SRCDIR)
$(SRCDIR): %: %.tar.gz
        tar -xmzf $<

.PHONY: start-spring-server
start-spring-server:
        $(PREFIX) ./spring.manifest -D -m /spring-rest/gradlew bootRun

clean: distclean

distclean:
        rm -rf $(SRCDIR)
````

When I run `make SGX=1` and then `make SGX_RUN=1` the Spring is built and I get the spring.manifest and spring.manifest.sgx executables, as well as spring.sig and spring.token files (see the output of make commands below):

````
$ make SGX=1

make SGX=1
ln -sf ../../../../../Runtime/pal_loader
tar -xmzf spring-rest.tar.gz
cd spring-rest && ./gradlew build
Starting a Gradle Daemon (subsequent builds will be faster)
:compileJava
:processResources
:classes
:findMainClass
:jar
:bootRepackage
:assemble
:compileTestJava
:processTestResources
:testClasses
:test
2018-04-11 12:25:06.270  INFO 4107 --- [       Thread-9] ationConfigEmbeddedWebApplicationContext : Closing org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@26365a5d: startup date [Wed Apr 11 12:25:05 WEST 2018]; root of context hierarchy
2018-04-11 12:25:06.270  INFO 4107 --- [       Thread-6] o.s.w.c.s.GenericWebApplicationContext   : Closing org.springframework.web.context.support.GenericWebApplicationContext@39d5cdef: startup date [Wed Apr 11 12:25:03 WEST 2018]; root of context hierarchy
:check
:build

BUILD SUCCESSFUL

Total time: 7.755 secs
[ spring.manifest ]
RELDIR= && \
sed -e 's:\$(PAL):/home/igor/graphene-for-pull-request/graphene/Runtime/pal_loader SGX:g' -e 's:\$(PWD):/home/igor/graphene-for-pull-request/graphene/LibOS/shim/test/apps/spring:g' -e 's:\$(BIN):spring:g' -e 's:\$(SHIMPATH):'$RELDIR'../../../../../Runtime/libsysdb.so:g' -e 's:\$(LIBCDIR):'$RELDIR'../../../../../Runtime:g' -e 's:\$(HOST):136.133.31.132:g' -e 's:\$(PORT):8090:g' spring.manifest.template > spring.manifest
(grep -q '^#!' spring.manifest && chmod +x spring.manifest) || true
for f in `grep -Po 'sgx.trusted_children.[^\\s=]+\\s*=\\s*file:\\K\\S+' spring.manifest`; do make ${f%.sig}.manifest.sgx; done
make[1]: Entering directory '/home/igor/graphene-for-pull-request/graphene/LibOS/shim/test/apps/spring'
make[1]: 'spring.manifest' is up to date.
make[1]: Leaving directory '/home/igor/graphene-for-pull-request/graphene/LibOS/shim/test/apps/spring'
[ spring.manifest.sgx ]
/home/igor/graphene-for-pull-request/graphene/Pal/src/host/Linux-SGX/signer/pal-sgx-sign -libpal /home/igor/graphene-for-pull-request/graphene/Pal/src/host/Linux-SGX/../../../../Runtime/libpal-Linux-SGX.so -key /home/igor/graphene-for-pull-request/graphene/Pal/src/host/Linux-SGX/signer/enclave-key.pem -output spring.manifest.sgx -manifest spring.manifest
Attributes:
    size:      268435456
    threadnum: 28
    isvprodid: 0
    isvsvn:    0
    flags:     0600000000000000
    xfrms:     0700000000000000
    miscs:     00000000
Trusted files:
    f5b3c6861a3e036b0585b71f800f557baabeeaf7e32853fd0b1a0fca1ed02571 file:/lib/x86_64-linux-gnu/libcrypto.so.1.0.0
    7749b38674433870a57cba74daa9d0c6206a8564ba9ece82b07d78acd426b2b3 file:/lib/x86_64-linux-gnu/libssl.so.1.0.0
    09edf863a352690b1ca8de2f8ec16c88389201dee721462d780efcfb88065b51 file:../../../../../Runtime/ld-linux-x86-64.so.2
    2c5907393d4231e9d91ebcd9f84b3519262bd39ef468b1278a8a3a76e10f7a50 file:../../../../../Runtime/libpthread.so.0
    650a2a0d1c979f0ed7df2915a5ba64b125908707fb41c2abe251240a3b18af5f file:../../../../../Runtime/libsysdb.so
    67f5815d370b12b5ab5130000c5c10420f86663c3cdcc8bf93fcb6e0b4faf401 file:../../../../../Runtime/libdl.so.2
    edcb489333b3aa9808f9bd84fe1c83a34c79ec4bf0698ceaa09839f44b74ae58 file:../../../../../Runtime/libm.so.6
    13d1eb5d8097409f53d616d1da31ec4f4f84cb492f33a5dfb55a0ec627ac1f22 spring-rest/gradlew
    0c73620c65da9d77ace4648c05741388e96245d6713cc802a4582c8336ad71e0 file:../../../../../Runtime/libc.so.6
    5c9b7c88dad7622cbc1eb09dbd71d7fec253d1cbd684ee848d84fded5cf43c86 file:spring-rest/gradle/wrapper/gradle-wrapper.jar
    de9531bf54031fb11bdfb7d6f3f8de998d5ef04252a0469de1fac4920ada94aa file:spring-rest/gradle/wrapper/gradle-wrapper.properties
    e62a9cecec95ba4741ee5cd08f33a266a1be8ef2d8533e9ab33a3a8c70761247 file:spring-rest/build.gradle
    11e338c599cb9f2aa6df5425cbf3c91fa9e0e1e390e179a04af35f586b267491 file:spring-rest/src/main/java/hello/storage/StorageFileNotFoundException.java
    a982a9e8b213631567de9253ae4bbb12f67d80b60abef0d4b76532aa4f686f40 file:spring-rest/src/main/java/hello/storage/StorageException.java
    4c02d06914a094f144f0660d5549cca0def2d8bb145ea88f995f1410f9e06415 file:spring-rest/src/main/java/hello/storage/StorageService.java
    fdd6bb4ad20eab447f2732343cf473f636eae7d90031d7f6d9aa2bc76f23c412 file:spring-rest/src/main/java/hello/storage/StorageProperties.java
    438a3263abe2749fcaec8d0cbce044828c7a4438757d8d22b080673f268eb91c file:spring-rest/src/main/java/hello/Application.java
    d8d8c206aab5fbe322b2797bb57c5cf63b4f51ed5fccc996ad2bde01f06b4b7b file:spring-rest/src/main/java/hello/storage/FileSystemStorageService.java
    02986618f5e0e110b2fe3a01cc43a43236750c1af714f82caa6fe05604219b04 file:spring-rest/src/main/java/hello/FileUploadController.java
    0dc1824e98ce9548605d2168bce93b61e86631f975048794a36a5730e35b8fe6 file:spring-rest/src/main/resources/templates/uploadForm.html
    04292c70febdcede1e787209475d0bd6aed44930cf6e1dc65691b18ee0a8fbde file:spring-rest/src/main/resources/application.properties
    578bc7d00ea6afd254166a6a09910232355285da1e4df85ee61b33e3b5c4dd98 file:spring-rest/src/test/java/hello/FileUploadIntegrationTests.java
    448745ea747c7750a7128431b893072f1171df7b8abd77d3a10be505893bb612 file:spring-rest/src/test/java/hello/storage/FileSystemStorageServiceTests.java
    fad95001fc23b35bfe74287991e61c0e78f7f039953e794bac168db00f6ce547 file:spring-rest/src/test/java/hello/FileUploadTests.java
    6e08f7d0fd517f4df55a15a4ff76b91e0c7d7a7df19dd8c3836f480b272dbd94 file:spring-rest/src/test/resources/hello/testupload.txt
Trusted children:
Memory:
    000000000fffe000-0000000010000000 [REG:R--] (manifest) measured
    000000000ffc5000-000000000fffd000 [REG:RW-] (ssa)
    000000000ffa8000-000000000ffc4000 [TCS:---] (tcs)
    000000000ff8b000-000000000ffa7000 [REG:RW-] (tls)
    000000000ff7a000-000000000ff8a000 [REG:RW-] (stack)
    000000000ff69000-000000000ff79000 [REG:RW-] (stack)
    000000000ff58000-000000000ff68000 [REG:RW-] (stack)
    000000000ff47000-000000000ff57000 [REG:RW-] (stack)
    000000000ff36000-000000000ff46000 [REG:RW-] (stack)
    000000000ff25000-000000000ff35000 [REG:RW-] (stack)
    000000000ff14000-000000000ff24000 [REG:RW-] (stack)
    000000000ff03000-000000000ff13000 [REG:RW-] (stack)
    000000000fef2000-000000000ff02000 [REG:RW-] (stack)
    000000000fee1000-000000000fef1000 [REG:RW-] (stack)
    000000000fed0000-000000000fee0000 [REG:RW-] (stack)
    000000000febf000-000000000fecf000 [REG:RW-] (stack)
    000000000feae000-000000000febe000 [REG:RW-] (stack)
    000000000fe9d000-000000000fead000 [REG:RW-] (stack)
    000000000fe8c000-000000000fe9c000 [REG:RW-] (stack)
    000000000fe7b000-000000000fe8b000 [REG:RW-] (stack)
    000000000fe6a000-000000000fe7a000 [REG:RW-] (stack)
    000000000fe59000-000000000fe69000 [REG:RW-] (stack)
    000000000fe48000-000000000fe58000 [REG:RW-] (stack)
    000000000fe37000-000000000fe47000 [REG:RW-] (stack)
    000000000fe26000-000000000fe36000 [REG:RW-] (stack)
    000000000fe15000-000000000fe25000 [REG:RW-] (stack)
    000000000fe04000-000000000fe14000 [REG:RW-] (stack)
    000000000fdf3000-000000000fe03000 [REG:RW-] (stack)
    000000000fde2000-000000000fdf2000 [REG:RW-] (stack)
    000000000fdd1000-000000000fde1000 [REG:RW-] (stack)
    000000000fdc0000-000000000fdd0000 [REG:RW-] (stack)
    000000000fdaf000-000000000fdbf000 [REG:RW-] (stack)
    000000000bb7f000-000000000bbaa000 [REG:R-X] (code) measured
    000000000bda9000-000000000bdab000 [REG:RW-] (data) measured
    000000000bdab000-000000000fdae000 [REG:RW-] (bss)
    000000000bb7c000-000000000bb7e000 [REG:-W-] (exec) measured
    0000000000000000-000000000bb7c000 [REG:RWX] (free)
Measurement:
    87659243284a2e8f08defa09ef278c3d08929e01780dac7092db79be59f882d2
````

and here is the output of `make SGX_RUN=1`

````
$ make SGX_RUN=1
tar -xmzf spring-rest.tar.gz
cd spring-rest && ./gradlew build
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:findMainClass
:jar
:bootRepackage
:assemble
:compileTestJava UP-TO-DATE
:processTestResources UP-TO-DATE
:testClasses UP-TO-DATE
:test UP-TO-DATE
:check UP-TO-DATE
:build

BUILD SUCCESSFUL

Total time: 0.962 secs
[ Token: spring ]
Attributes:
    mrenclave: 87659243284a2e8f08defa09ef278c3d08929e01780dac7092db79be59f882d2
    isvprodid: 0
    isvsvn:    0
    flags:     0600000000000000
    xfrms:     0700000000000000
    miscs:     00000000
    modulus:   b9b27940a8aa0cefd25ddda23c37be36...
    exponent:  3
    signature: 0acc21e4cf937688ea7c104423bacb63...
````

The problem starts when I try to start the Spring Server with the `make start-spring-server` command:

````
$ make start-spring-server 
./spring.manifest -D -m /spring-rest/gradlew bootRun
can't read manifest: equal mark expected
invalid manifest: file:spring.manifest.sgx
Makefile:30: recipe for target 'start-spring-server' failed
make: *** [start-spring-server] Error 234
````

I think I'm missing something, but I can't seem to find what exactly. The syntax of the manifest file seems to be valid, so I'm not sure what's wrong. The error message comes from the `Pal/lib/graphene/config.c` file that reads the manifest file. Seems like it fails on some line.

**Comment by thomasknauth** [comment]
Try removing any non-alphabetic/numeric characters from manifest identifiers. Instead of sgx.trusted_files.test-java1 try testjava1. The parsing logic for the manifest is rather strict sometimes.

**Comment by chiache** [comment]
I believe the problem is caused by the hyphens (-) in the configuration keys (the part before the equal mark). For the values (the part after the equal mark), using hyphens is fine.

Apology for failing to document this limitation, and the unhelpful error message. I will create another PR to clarify it.

**Comment by zavalyshyn** [comment]
Great. Indeed the problem was there. Thanks for clarification and help

**Comment by masomel** [comment]
Not sure if there's a PR for this issue yet. I just encountered the same error message when my Makefile (copied from the bash example) generated the trusted library entries with ldd, which in my application included an entry for `sgx.trusted_files.libstdc++`. So it'd be awesome if the documentation could include such cases, or eventually have support for "+" in configuration keys as well.

**Comment by dimakuv** [comment]
@masomel There is no PR for this yet. We need to replace the whole manifest syntax with something well-known and production-quality like JSON (rather than patching corner cases like `+` and `-` symbols). Sorry for this confusion.

**Comment by dimakuv** [comment]
@masomel That said, I updated our Wiki to reflect that only alphanumeric characters in keys are supported. See https://github.com/oscarlab/graphene/wiki/Graphene-SGX-Manifest-Syntax and https://github.com/oscarlab/graphene/wiki/Graphene-Manifest-Syntax.

**Comment by mkow** [comment]
btw. The issue about switching to a different manifest syntax: https://github.com/oscarlab/graphene/issues/870
I think right now the plan is to try TOML.

**Comment by masomel** [comment]
@dimakuv Thanks for updating the documentation! I do want to add, though, that developers should be reminded in general that modifying the generated manifest afterward it has been signed will cause the enclave measurement verification to fail when Graphene loads the application. This wasn't immediately obvious to me when I modified the auto-generated manifest to solve this issue. As far as I could tell, this security property isn't well-documented outside of the paper. 

Maybe my particular case isn't super common, but it seems like you're generally moving in the direction of using ldd to capture all required libraries, so until the syntax is updated, you may want to at least notify developers that any modifications to fix unsupported characters need to happen before the manifest is signed.

**Comment by dimakuv** [comment]
@masomel This will be fixed by https://github.com/oscarlab/graphene/pull/1204.

**Comment by mkow** [comment]
Closing this issue, we don't plan to fix up the current manifest syntax/parsing implementation, but we'll switch to something totally different instead (see #870 to track progress/discuss this change).
